### ISSUE TITLE
Cold Start Forecasting

### ISSUE THREAD

**Issue by StatMixedML** [awslabs/gluon-ts#703]
## Description

This issue is closely related to https://github.com/awslabs/gluon-ts/issues/667#issue-570373860 and https://github.com/awslabs/gluon-ts/issues/684#issue-572827024. I am trying to mimic a cold-start problem using DeepAR. As such, the field *target* of test data set needs to be [], as the target has never been observed. I am using a toy data set, where the *train_data* would look like

```python
{'target': array([10., 12., 16., 15., 16., 20., 21., 19., 20., 19., 14., 13., 12.,
         9.], dtype=float32),
 'feat_static_cat': array([1]),
 'start': Timestamp('2015-01-04 00:00:00', freq='W-SUN')}
```

and *test_data* would look like

```python
{'target': array([], dtype=float32),
 'feat_static_cat': array([1]),
 'start': Timestamp('2018-12-30 00:00:00', freq='W-SUN')}
```
I also found an aws [tutorial ](https://github.com/awslabs/amazon-sagemaker-examples/blob/27b6055543c33bda7e6ec5a323d17e11d5f3b83f/introduction_to_amazon_algorithms/deepar-retail-forecasting/deepar_retail_forecasting.ipynb) that shows at the very bottom the following

```python
product_category = 'shoe'

cat = int(le.transform([product_category])[0])

request_json = {
 "instances": [
  {"start": "2019-06-12 00:00:00", "target": [], "cat": cat} # No target values because this is a new product with no existing time series sales data
 ],
 "configuration": {
  "num_samples": 10,
  "output_types": ["quantiles"],
  "quantiles": ["0.9"]
 }
}
```
saying that, indeed, the target needs to be empty. The model is estimated in a standard fashion

```python
estimator = DeepAREstimator(
    freq="W",
    prediction_length=12,
    distr_output=StudentTOutput(),
    trainer=Trainer(epochs=20, learning_rate=0.01),
    use_feat_static_cat=True,
    cardinality=[4]
)
predictor = estimator.train(training_data=train_data)
```

 All is working decently well if I use 

```python
predictor.predict(test_data)
```

However, when I use 

```python
forecast_it, ts_it = make_evaluation_predictions(dataset = test_data,      
                                                 predictor = predictor,  
                                                 num_samples = 100)
forecasts = list(forecast_it)
```

I get the follwoing error message

```python
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-44-9071bd2d1731> in <module>
      2                                                  predictor = predictor,  # predictor
      3                                                  num_samples = 100)
----> 4 forecasts = list(forecast_it)

~\.conda\envs\gluonts\lib\site-packages\gluonts\model\predictor.py in predict(self, dataset, num_samples)
    307             freq=self.freq,
    308             output_transform=self.output_transform,
--> 309             num_samples=num_samples,
    310         )
    311 

~\.conda\envs\gluonts\lib\site-packages\gluonts\model\forecast_generator.py in __call__(self, inference_data_loader, prediction_net, input_names, freq, output_transform, num_samples, **kwargs)
    193         **kwargs
    194     ) -> Iterator[Forecast]:
--> 195         for batch in inference_data_loader:
    196             inputs = [batch[k] for k in input_names]
    197             outputs = prediction_net(*inputs).asnumpy()

~\.conda\envs\gluonts\lib\site-packages\gluonts\dataset\loader.py in __iter__(self)
    261     def __iter__(self) -> Iterator[DataBatch]:
    262         buffer = BatchBuffer(self.batch_size, self.ctx, self.dtype)
--> 263         for data_entry in self.transform(iter(self.dataset), is_train=False):
    264             buffer.add(data_entry)
    265             if len(buffer) >= self.batch_size:

~\.conda\envs\gluonts\lib\site-packages\gluonts\transform\_base.py in __call__(self, data_it, is_train)
    139     ) -> Iterator:
    140         num_idle_transforms = 0
--> 141         for data_entry in data_it:
    142             num_idle_transforms += 1
    143             try:

~\.conda\envs\gluonts\lib\site-packages\gluonts\transform\_base.py in __call__(self, data_it, is_train)
     90         self, data_it: Iterator[DataEntry], is_train: bool
     91     ) -> Iterator:
---> 92         for data_entry in data_it:
     93             try:
     94                 yield self.map_transform(data_entry.copy(), is_train)

~\.conda\envs\gluonts\lib\site-packages\gluonts\transform\_base.py in __call__(self, data_it, is_train)
     90         self, data_it: Iterator[DataEntry], is_train: bool
     91     ) -> Iterator:
---> 92         for data_entry in data_it:
     93             try:
     94                 yield self.map_transform(data_entry.copy(), is_train)

~\.conda\envs\gluonts\lib\site-packages\gluonts\transform\_base.py in __call__(self, data_it, is_train)
     90         self, data_it: Iterator[DataEntry], is_train: bool
     91     ) -> Iterator:
---> 92         for data_entry in data_it:
     93             try:
     94                 yield self.map_transform(data_entry.copy(), is_train)

~\.conda\envs\gluonts\lib\site-packages\gluonts\transform\_base.py in __call__(self, data_it, is_train)
     90         self, data_it: Iterator[DataEntry], is_train: bool
     91     ) -> Iterator:
---> 92         for data_entry in data_it:
     93             try:
     94                 yield self.map_transform(data_entry.copy(), is_train)

~\.conda\envs\gluonts\lib\site-packages\gluonts\transform\_base.py in __call__(self, data_it, is_train)
     90         self, data_it: Iterator[DataEntry], is_train: bool
     91     ) -> Iterator:
---> 92         for data_entry in data_it:
     93             try:
     94                 yield self.map_transform(data_entry.copy(), is_train)

~\.conda\envs\gluonts\lib\site-packages\gluonts\transform\_base.py in __call__(self, data_it, is_train)
     90         self, data_it: Iterator[DataEntry], is_train: bool
     91     ) -> Iterator:
---> 92         for data_entry in data_it:
     93             try:
     94                 yield self.map_transform(data_entry.copy(), is_train)

~\.conda\envs\gluonts\lib\site-packages\gluonts\transform\_base.py in __call__(self, data_it, is_train)
     90         self, data_it: Iterator[DataEntry], is_train: bool
     91     ) -> Iterator:
---> 92         for data_entry in data_it:
     93             try:
     94                 yield self.map_transform(data_entry.copy(), is_train)

~\.conda\envs\gluonts\lib\site-packages\gluonts\transform\_base.py in __call__(self, data_it, is_train)
     90         self, data_it: Iterator[DataEntry], is_train: bool
     91     ) -> Iterator:
---> 92         for data_entry in data_it:
     93             try:
     94                 yield self.map_transform(data_entry.copy(), is_train)

~\.conda\envs\gluonts\lib\site-packages\gluonts\transform\_base.py in __call__(self, data_it, is_train)
     90         self, data_it: Iterator[DataEntry], is_train: bool
     91     ) -> Iterator:
---> 92         for data_entry in data_it:
     93             try:
     94                 yield self.map_transform(data_entry.copy(), is_train)

~\.conda\envs\gluonts\lib\site-packages\gluonts\transform\dataset.py in __iter__(self)
     42 
     43     def __iter__(self) -> Iterator[DataEntry]:
---> 44         yield from self.transformations(self.base_dataset, is_train=True)
     45 
     46     def __len__(self):

~\.conda\envs\gluonts\lib\site-packages\gluonts\transform\_base.py in __call__(self, data_it, is_train)
     94                 yield self.map_transform(data_entry.copy(), is_train)
     95             except Exception as e:
---> 96                 raise e
     97 
     98     @abc.abstractmethod

~\.conda\envs\gluonts\lib\site-packages\gluonts\transform\_base.py in __call__(self, data_it, is_train)
     92         for data_entry in data_it:
     93             try:
---> 94                 yield self.map_transform(data_entry.copy(), is_train)
     95             except Exception as e:
     96                 raise e

~\.conda\envs\gluonts\lib\site-packages\gluonts\transform\_base.py in map_transform(self, data, is_train)
    107 
    108     def map_transform(self, data: DataEntry, is_train: bool) -> DataEntry:
--> 109         return self.transform(data)
    110 
    111     @abc.abstractmethod

~\.conda\envs\gluonts\lib\site-packages\gluonts\transform\_base.py in transform(self, data)
    126 
    127     def transform(self, data: DataEntry) -> DataEntry:
--> 128         return self.func(data.copy())
    129 
    130 

~\.conda\envs\gluonts\lib\site-packages\gluonts\evaluation\backtest.py in truncate_target(data)
     87         target = data["target"]
     88         assert (
---> 89             target.shape[-1] >= prediction_length
     90         )  # handles multivariate case (target_dim, history_length)
     91         data["target"] = target[..., :-prediction_length]

AssertionError: 
```
The crucial part is this

```python
     91         data["target"] = target[..., :-prediction_length]
```
as it tries to remove the *prediction_length* part from the data set, as *test_set ~ train_set + prediction_length*, as described [here](https://gluon-ts.mxnet.io/examples/extended_forecasting_tutorial/extended_tutorial.html#1.1.1-What-is-in-a-dataset?)

> dataset.test is an iterable collection of data entries used for inference. The test dataset is an extended version of the train dataset that contains a window in the end of each time series that was not seen during training. This window has length equal to the recommended prediction length.

## Help
So I am asking of how to 

- properly prepare a train and test set so as to produce cold-start forecasts
- properly use the prediction arguments

Ideally, we would **have an official notebook that shows each step in detail**, so as to avoid confusion on how to use DeepAR for forecasting cold starts, as it has already been raised numerous times.

Many thanks!

## Environment
- Operating system: Windows
- Python version: 3.6.7
- GluonTS version: 0.4.3.dev55+g90587ba

**Comment by lostella** [comment]
@StatMixedML the make_evaluation_predictions function is used essentially to create forecast/ground truth pairs that can be use to measure forecasting accuracy later on. This is why it needs to remove the trailing prediction_length observation from the series.

If there is no data in the series to compare against, this is not possible. To evaluate cold start forecasting accuracy you will need to provide series with prediction_length observations: these will be kept in the ground truth, but removed from the conditioning range (ie what the models gets as input).

To summarize: producing cold start forecasts just amounts to giving an empty target to the predictor, as you say; to *evaluate* cold start forecasts, you need to provide prediction_length observations in the target.

**Comment by StatMixedML** [comment]
@lostella Thanks for clarifying!
